eng/scripts/live-test-resource-cleanup.ps1

```diff
@@ -185,30 +187,39 @@ function DeleteOrUpdateResourceGroups() {
   }
 }
 
-if (!(Get-AzContext) -and !$Login) {
-  Write-Error "Caller must be logged in via `Connect-AzAccount` or -Login parameter must be set."
-  exit 1
-}
-
-if ($Login) {
-  Write-Verbose "Logging in"
-
+function Login() {
   if ($PSCmdlet.ParameterSetName -eq "Provisioner") {
+    Write-Verbose "Logging in with provisioner"
     $provisionerSecret = ConvertTo-SecureString -String $ProvisionerApplicationSecret -AsPlainText -Force
     $provisionerCredential = [System.Management.Automation.PSCredential]::new($ProvisionerApplicationId, $provisionerSecret)
     Connect-AzAccount -Force -Tenant $TenantId -Credential $provisionerCredential -ServicePrincipal -Environment $Environment
-  } else {
+    Select-AzSubscription -Subscription $SubscriptionId -Confirm:$false
+  } elseif ($Login) {
+    Write-Verbose "Logging in with interactive user"
     Connect-AzAccount
+  } elseif (Get-AzContext) {
+    Write-Verbose "Using existing account"
+  } else {
+    $errMsg = 'User context not found. Please re-run script with "-Login" to login, ' +
+              'or run "Connect-AzAccount -UseDeviceAuthentication" if interactive login is not available.'
+    Write-Error $errMsg
+    exit 1
   }
 }
 
-$originalSubscription = (Get-AzContext).Subscription.Id
-if ($SubscriptionId -and ($originalSubscription -ne $SubscriptionId)) {
-  Select-AzSubscription -Subscription $SubscriptionId -Confirm:$false
-}
+function Main() {
+  Login
+
+  $originalSubscription = (Get-AzContext).Subscription.Id
+  if ($SubscriptionId -and ($originalSubscription -ne $SubscriptionId)) {
+    Select-AzSubscription -Subscription $SubscriptionId -Confirm:$false
+  }
 
-DeleteOrUpdateResourceGroups
+  DeleteOrUpdateResourceGroups
 
-if ($SubscriptionId -and ($originalSubscription -ne $SubscriptionId)) {
-  Select-AzSubscription -Subscription $originalSubscription -Confirm:$false
+  if ($SubscriptionId -and ($originalSubscription -ne $SubscriptionId)) {
+    Select-AzSubscription -Subscription $originalSubscription -Confirm:$false
+  }
 }
+
+Main
```

Comment:
The reason I tend to avoid that approach is many of our scripts start out small with code in the top level context. Over time people add incremental updates to them, and a year or two later we have a large spaghetti code script that requires refactoring in order to re-use functionality.

Perhaps `Main` is an odd name here though, vs. something like `CleanupResources`?

Reply:
The name wasn't an issue. Main is fine. It was more that it wasn't necessary. I appreciate your point, however. Personally, I solve this in my own scripts by putting all supporting functions at the bottom. With my original test resources scripts I put them at the top because I was afraid those reviewing who didn't know powershell as well would not realize the ordered parse doesn't affect execution in these cases.

Reply:
I usually put my logic on the bottom but I like your approach of putting them at the top instead. I think there's a happy medium here of slimming down the execution code vs. having it all in a Main function.

Reply:
Whatever you want to do. I prefaced it with "tip" (was nearly a "nit") since it's not blocking. 🙂

Reply:
@heaths I was trying it out, it appears powershell doesn't do forward declaration, so now I'm not sure what you were describing?

Reply:
Sorry, you're right. In the times I (obviously incorrectly) recalled, I must used them in a child scope (function), as this shows working:

```powershell
[CmdletBinding()]
param ()

function LogError([string] $Message) {
    Log $Message 'Red'
}

function Log([string] $Message, [ConsoleColor] $ForegroundColor = [ConsoleColor]::White) {
    Write-Host $Message -Foreground $ForegroundColor
}

Log "Starting something..."
LogError "Uh, oh!"
```

Moving `LogError` after the "main" code errs. That said - and again it's only a nit - putting your `Main` at the end and then calling it doesn't really help the situation of making it obvious. Now, perhaps if you declared `Main` at the top and helper functions after that, then called `Main`, it does put the "main code" front-and-center. Just a thought.

Reply:
Think I would do away with the Main method as just put the script at the bottom. That is what we have done in most of our other scripts that aren't just helpers.

Reply:
What if I created a sub cleanup helper file, and then a much smaller script containing the script scope logic? This way we still follow the simpler cmdlet pattern but without making it hard to update code in the future to share logic, add tests, etc.?